<center>
<h1> TP-Projet d'optimisation numérique </h1>
<h1> Année 2020-2021 - 2e année département Sciences du Numérique </h1>
<h1> Noms: Sajid - Wissad </h1>
<h1> Prénoms: Badr - Mehdi </h1>    
</center>

# Algorithme de Newton
## Implémentation 
 
1. Coder l’algorithme de Newton local tel que décrit dans la section *Algorithme de Newton* (fichier `Algorithme_De_Newton.jl`)

2. Tester l’algorithme sur les fonctions $f_{1}$ , $f_{2}$ avec les points initiaux $x_{011}$ , $x_{012}$ (pour $f_{1}$ ) et $x_{021}$ , $x_{022}$ , $x_{023}$ (pour $f_{2}$ ) donnés en Annexe A.

In [1]:
#using Pkg; Pkg.add("LinearAlgebra"); Pkg.add("Markdown")
# using Documenter
using LinearAlgebra
using Markdown                             # Pour que les docstrings en début des fonctions ne posent
                                           # pas de soucis. Ces docstrings sont utiles pour générer 
                                           # la documentation sous GitHub
include("Algorithme_De_Newton.jl")

# Affichage les sorties de l'algorithme des Régions de confiance
function my_afficher_resultats(algo,nom_fct,point_init,xmin,fxmin,flag,sol_exacte,nbiters)
	println("-------------------------------------------------------------------------")
	printstyled("Résultats de : ",algo, " appliqué à ",nom_fct, " au point initial ", point_init, ":\n",bold=true,color=:blue)
	println("  * xsol = ",xmin)
	println("  * f(xsol) = ",fxmin)
	println("  * nb_iters = ",nbiters)
	println("  * flag = ",flag)
	println("  * sol_exacte : ", sol_exacte)
end

# Fonction f0
# -----------
f0(x) =  sin(x)
# la gradient de la fonction f0
grad_f0(x) = cos(x)
# la hessienne de la fonction f0
hess_f0(x) = -sin(x)
sol_exacte = -pi/2
options = []

x0 = sol_exacte
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = -pi/2+0.5
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = pi/2
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f0,grad_f0,hess_f0,x0,options)
my_afficher_resultats("Newton","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)


# Fonction f1
# -----------
f1(x) = 2*(x[1]+x[2]+x[3]-3)^2 + (x[1]-x[2])^2 + (x[2]-x[3])^2
# la gradient de la fonction f1
grad_f1(x) = [4*(x[1]+x[2]+x[3]-3) + 2*(x[1]-x[2]); 4*(x[1]+x[2]+x[3]-3) - 2*(x[1]-x[2])+2*(x[2]-x[3]); 4*(x[1]+x[2]+x[3]-3)-2*(x[2]-x[3])]
# la hessienne de la fonction f1
hess_f1(x) = [6 2 4; 2 8 2; 4 2 6]
sol_exacte = [1,1,1]
options = []

x0 = [1; 0; 0]
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f1,grad_f1,hess_f1,x0,options)
my_afficher_resultats("Newton","f1",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [10; 3; -2.2]
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f1,grad_f1,hess_f1,x0,options)
my_afficher_resultats("Newton","f1",x0,xmin,f_min,flag,sol_exacte,nb_iters)


# Fonction f2
# -----------
f2(x) = 100*(x[2]-x[1]^2)^2+(1-x[1])^2
grad_f2(x) = [-400*x[1]*(x[2]-x[1]^2)-2*(1-x[1]) ; 200*(x[2]-x[1]^2)]
hess_f2(x) = [-400*(x[2]-3*x[1]^2)+2  -400*x[1]; -400*x[1]  200]
sol_exacte = [1,1,1]
options = []

x0 = [-1.2; 1]
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("Newton","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [10; 0]
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("Newton","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [0; ((1/200)+(1/(10^12)))]
xmin,f_min,flag,nb_iters = Algorithme_De_Newton(f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("Newton","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)


-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.5707963267948966:
  * xsol = -1.5707963267948966
  * f(xsol) = -1.0
  * nb_iters = 1
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial -1.0707963267948966:
  * xsol = -1.5707963267949088
  * f(xsol) = -1.0
  * nb_iters = 3
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f0 au point initial 1.5707963267948966:
  * xsol = 1.5707963267948966
  * f(xsol) = 1.0
  * nb_iters = 1
  * flag = 0
  * sol_exacte : -1.5707963267948966
-------------------------------------------------------------------------
Résultats de : Newton appliqué à f1 au point initial [1, 0, 0]:
  * xsol = [1.0, 1.0, 0.9999999999999999]
  * f(xsol) = 1.232595164407

LoadError: SingularException(2)

## Interprétation 

Justifier

1. les résultats obtenus pour l'exemple $f_0$ ci-dessus;

2. que l’algorithme implémenté converge en une itération pour $f_{1}$;

3. que l’algorithme puisse ne pas converger pour $f_{2}$ avec certains points initiaux.

## Vos réponses?
justification : 

1.Dans les deux premiers cas avec les deux états initiaux suivant :  -1.5707963267948966 , -1.0707963267948966
on a une conververgence vers le minimum. et ce résultat est cohèrent car c'est bien le minimum du sinus.
Par contre dans le troisième cas avec le point : 1.5707963267948966 on n'a pas de convergence car dans notre approximatino de Taylor à l'ordre deux nous donne un fonction qui au fil des itérations se dirige vers le sens oppose a un minimum, le résultat nous donne bien le maximum de la fonction sinus.   

2.Puisque $f_{1}$ est égale à son dévelopement de Tailor à l'ordre 2, une seul itération suffit pour atteindre le minimum de $f_{1}$ et résoudre le système lineaire qui lui est associé.

3.La solution peut ne pas converger, car le système lineaire  peut ne pas avoir de solution dans le cas où  $\nabla^{2} f (x_{k})$ n'est pas inversible , et cela peut venir de notre approximation initial qui nous mene a une fonction qui au fil des itérations se dirige vers le sens opposé a son minimum.

# Régions de confiance avec pas de cauchy 

## Implémentation 

1. Coder l'algorithme du pas de Cauchy d’un sous-problème de
régions de confiance (fichier `Pas_De_Cauchy.jl`). Tester sur les quadratiques proposées en Annexe B.

2. Coder l'algorithme de régions de confiance (fichier `Regions_De_Confiance.jl`). Tester sur les problèmes de l’Annexe A.

In [12]:
include("Pas_De_Cauchy.jl")

"B. Cas tests pour le calcul du pas de Cauchy"

"Quadratique 1"
g1 = [0; 0]
H1 = [7 0; 0 2]
s,e = Pas_De_Cauchy(g1,H1,1)
println("Cauchy appliqué sur la quadratique 1\nL'approximation de la solution est : s = ", s,"\n")


"Quadratique 2"
g2 = [6; 2]
H2 = [7 0; 0 2]
s,e = Pas_De_Cauchy(g2,H2,1)
println("Cauchy appliqué sur la quadratique 2\nL'approximation de la solution est : s = ", s,"\n")

"Quadratique 3"
g3 = [-2; 1]
H3 = [-2 0; 0 10]
s,e = Pas_De_Cauchy(g3,H3,1)
println("Cauchy appliqué sur la quadratique 3\nL'approximation de la solution est : s = ", s,"\n")


Cauchy appliqué sur la quadratique 1
L'approximation de la solution est : s = [0, 0]

Cauchy appliqué sur la quadratique 2
L'approximation de la solution est : s = [-0.9230769230769234, -0.30769230769230776]

Cauchy appliqué sur la quadratique 3
L'approximation de la solution est : s = [0.8944271909999159, -0.4472135954999579]



In [71]:
# Vos tests
include("Regions_De_Confiance.jl")
include("Pas_De_Cauchy.jl")
include("Gradient_Conjugue_Tronque.jl")


# Fonction f0
# -----------
#sol_exacte = -pi/2
#options = []

#x0 = sol_exacte
#xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f0,grad_f0,hess_f0,x0,options)
#my_afficher_resultats("cauchy","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
#x0 = -pi/2+0.5
#xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f0,grad_f0,hess_f0,x0,options)
#my_afficher_resultats("cauchy","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
#x0 = pi/2
#xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f0,grad_f0,hess_f0,x0,options)
#my_afficher_resultats("cauchy","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)


# Fonction f1
# -----------
sol_exacte = [1,1,1]
options = []

x0 = [1; 0; 0]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f1,grad_f1,hess_f1,x0,options)
my_afficher_resultats("cauchy","f1",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [10; 3; -2.2]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f1,grad_f1,hess_f1,x0,options)
my_afficher_resultats("cauchy","f1",x0,xmin,f_min,flag,sol_exacte,nb_iters)


# Fonction f2
# -----------
sol_exacte = [1,1,1]
options = []

x0 = [-1.2; 1]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("cauchy","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [10; 0]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("cauchy","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [0; ((1/200)+(1/(10^12)))]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("cauchy",f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("cauchy","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)

La decroissance est : 8.532423208191126
La decroissance est : 0.33595195967772484
La decroissance est : 0.07344828445539156
La decroissance est : 0.028791670087372476
La decroissance est : 0.014519505968362448
La decroissance est : 0.007344121438614328
La decroissance est : 0.0037158119564451465
La decroissance est : 0.0018800453861078804
La decroissance est : 0.0009512244601960222
La decroissance est : 0.0004812798569662275
La decroissance est : 0.0002435075110393692
-------------------------------------------------------------------------
Résultats de : cauchy appliqué à f1 au point initial [1, 0, 0]:
  * xsol = [1.0000558873349883, 0.999992420017735, 0.9999289527004819]
  * f(xsol) = 9.090411079109608e-9
  * nb_iters = 26
  * flag = 2
  * sol_exacte : [1, 1, 1]
La decroissance est : 91.70805308544246
La decroissance est : 63.06553314849053
La decroissance est : 25.205923828703913
La decroissance est : 10.359280377487256
La decroissance est : 4.300145725758862
La decroissance est : 1

## Interprétation 

1. Quelle relation lie la fonction test $f_1$ et son modèle de Taylor à l’ordre 2 ? Comparer alors les performances de Newton et RC-Pas de Cauchy sur cette fonction.

2. Le rayon initial de la région de confiance est un paramètre important dans l’analyse
de la performance de l’algorithme. Sur quel(s) autre(s) paramètre(s) peut-on jouer
pour essayer d’améliorer cette performance ? Étudier l’influence d’au moins deux de
ces paramètres.

## Vos réponses?


1. $f_{1}$ est égale à son dévelopement de Tailor à l'ordre 2, avec l'algorithme de Newton la convergence se fait en une seul itèration, car on minimise dans ce cas directement la fonction quadratique. À cause des différentes assertions au début de la mèthode de Cauchy, Pour um même minimum local, il nous faut plusieur pas en avant, contrairement à l'algorithme de Newton qui dans ce cas est plus efficace,Or RC-Pas de Cauchy sera meilleurs dans le cas où on a des fonctions plus complexes. 

2. Tester Pour essayer d’améliorer cette performance on suppose de changer les paramètres suivant:

$\qquad$ $\gamma_{1}$ et $\gamma_{2}$ qui sont les facteurs d'agradissement et de réducation de la région de confiance . 

$\qquad$ $\eta_{1}$ et $\eta_{2}$ qui sont les critères d'agradissement et de réducation de la région de confiance . 

$\qquad$ $\Delta_{max}$ qui est le rayon de confiance maximal.

Si on change $\eta_1$ et $\eta_2$  avec  $\eta_1 < \eta_2 < 1$,  on constate que si on diminue l'intervalle (on diminue la chance de  retrouver $\rho_k$ dans cet intervalle),pour minimiser la fonctions, ainsi dans ce cas on a besoin de plusieurs itérations. On fait que résoudre le modèle 2.1 avec un $\Delta_k$ plus petit. 

# Régions de confiance avec gradient conjugué tronqué

## Implémentation 

1. Implémenter l’algorithme du Gradient Conjugué Tronqué, en se basant sur le cours (fichier `Gradient_Conjugue_Tronque.jl`).
On validera les résultats sur les fonctions de l’Annexe C.

2. Intégrer finalement l’algorithme du Gradient Conjugué Tronqué dans le code de
régions de confiance, et appliquer ce code pour résoudre les exemples proposés en
Annexe A.

In [14]:
include("Gradient_Conjugue_Tronque.jl")

"C. Cas tests pour la résolution du sous-problème par l’algorithme du Gradient Conjugué Tronqué"

"Quadratique 4"
g4 = [0; 0]
H4 = [-2 0; 0 10]
s = Gradient_Conjugue_Tronque(g4,H4,[])
println("gct appliqué sur la quadratique 4\nLe pas qui approche la solution est : s = ",s,"\n")


"Quadratique 5"
g5 = [2; 3]
H5 = [4 6; 6 5]
s = Gradient_Conjugue_Tronque(g5,H5,[])
println("gct appliqué sur la quadratique 5\nLe pas qui approche la solution est : s = ",s,"\n")

"Quadratique 6"
g6 = [2; 0]
H6 = [4 0; 0 -15]
s = Gradient_Conjugue_Tronque(g6,H6,[])
println("gct appliqué sur la quadratique 6\nLe pas qui approche la solution est : s = ",s,"\n")

gct appliqué sur la quadratique 4
Le pas qui approche la solution est : s = [0.0, 0.0]

gct appliqué sur la quadratique 5
Le pas qui approche la solution est : s = [1.1782448197996298, -1.6160876042514951]

gct appliqué sur la quadratique 6
Le pas qui approche la solution est : s = [-0.5, 0.0]



In [72]:
# Vos tests
include("Regions_De_Confiance.jl")
include("Pas_De_Cauchy.jl")
include("Gradient_Conjugue_Tronque.jl")

# Fonction f0
# -----------
#sol_exacte = -pi/2
#options = []

#x0 = sol_exacte
#xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f0,grad_f0,hess_f0,x0,options)
#my_afficher_resultats("gct","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
#x0 = -pi/2+0.5
#xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f0,grad_f0,hess_f0,x0,options)
#my_afficher_resultats("gct","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)
#x0 = pi/2
#xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f0,grad_f0,hess_f0,x0,options)
#my_afficher_resultats("gct","f0",x0,xmin,f_min,flag,sol_exacte,nb_iters)



# Fonction f1
# -----------
sol_exacte = [1,1,1]
options = []

x0 = [1; 0; 0]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f1,grad_f1,hess_f1,x0,options)
my_afficher_resultats("gct","f1",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [10; 3; -2.2]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f1,grad_f1,hess_f1,x0,options)
my_afficher_resultats("gct","f1",x0,xmin,f_min,flag,sol_exacte,nb_iters)


# Fonction f2
# -----------
sol_exacte = [1,1,1]
options = []

x0 = [-1.2; 1]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("gct","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [10; 0]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("gct","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)
x0 = [0; ((1/200)+(1/(10^12)))]
xmin,f_min,flag,nb_iters = Regions_De_Confiance("gct",f2,grad_f2,hess_f2,x0,options)
my_afficher_resultats("gct","f2",x0,xmin,f_min,flag,sol_exacte,nb_iters)

La decroissance est : 0.0
-------------------------------------------------------------------------
Résultats de : gct appliqué à f1 au point initial [1, 0, 0]:
  * xsol = [1, 0, 0]
  * f(xsol) = 9
  * nb_iters = 1
  * flag = 1
  * sol_exacte : [1, 1, 1]
La decroissance est : 91.70805308544246
La decroissance est : 0.0
-------------------------------------------------------------------------
Résultats de : gct appliqué à f1 au point initial [10.0, 3.0, -2.2]:
  * xsol = [8.411206090525074, 2.029851506603806, -2.9311264008203204]
  * f(xsol) = 106.01194691455754
  * nb_iters = 2
  * flag = 1
  * sol_exacte : [1, 1, 1]
La decroissance est : 0.0
-------------------------------------------------------------------------
Résultats de : gct appliqué à f2 au point initial [-1.2, 1.0]:
  * xsol = [-1.2, 1.0]
  * f(xsol) = 24.199999999999996
  * nb_iters = 1
  * flag = 1
  * sol_exacte : [1, 1, 1]
La decroissance est : 591198.8755806963
La decroissance est : 0.0
---------------------------------

## Interprétation  

1. Comparer la décroissance obtenue avec celle du pas de Cauchy, en retournant, dans
un premier temps le dernier itéré admissible à courbure positive (c’est à dire, que si
l’une ou l’autre des deux conditions (b) ou (d) sont rencontrées dans l’algorithme 3,
alors on ne calcule pas ``σ_{j}`` et on retourne le dernier itéré ``s_{j}`` directement).

2. Comparer la décroissance obtenue avec celle du pas de Cauchy, en imposant la sortie
dans l’algorithme 3 au bout d’une itération seulement. Que remarquez vous ?

3. Comparer la décroissance obtenue avec celle du pas de Cauchy dans le cas général.

4. Quels sont les avantages et inconvénients des deux approches ?

## Vos réponses?


En utilisant la fonction f2 au point initial [10, 0] : 
1. En retournant le dernier itéré admissible à courbure positive, le résultat de la décroissance obtenue est le suivant :

$\qquad$Pas de cauchy : $\qquad$$\qquad$ $\qquad$ $\qquad$ gct :
* $\qquad$ 591198.87 $\qquad$$\qquad$$\qquad$$\qquad$ 591198.87
* $\qquad$ 328991.13 $\qquad$$\qquad$$\qquad$$\qquad$ 337809.30
* $\qquad$ 64598.674 $\qquad$$\qquad$$\qquad$$\qquad$ 65927.96
* $\qquad$ 12522.069 $\qquad$$\qquad$$\qquad$$\qquad$ 5134.153
* $\qquad$ 2336.1646 $\qquad$$\qquad$$\qquad$$\qquad$ 1.160071

On remarque que la décroissance de l'algorithme "Pas de cauchy" et celle de l'algorithme "gct" sont à peu près Égaux.

2. En imposant la sortie dans l’algorithme "gct" au bout d’une itération seulement, le résultat de la décroissance obtenue est le suivant :

$\qquad$Pas de cauchy : $\qquad$$\qquad$ $\qquad$ $\qquad$ gct :
* $\qquad$ 591198.87 $\qquad$$\qquad$$\qquad$$\qquad$ 591198.87
* $\qquad$ 328991.13 $\qquad$$\qquad$$\qquad$$\qquad$ 337809.30
* $\qquad$ 64598.674 $\qquad$$\qquad$$\qquad$$\qquad$ 65927.96
* $\qquad$ 12522.069 $\qquad$$\qquad$$\qquad$$\qquad$ 5134.153
* $\qquad$ 2336.1646 $\qquad$$\qquad$$\qquad$$\qquad$ 1.160071



3. Dans le cas général, le résultat de la décroissance obtenue est le suivant :

$\qquad$Pas de cauchy : $\qquad$$\qquad$ $\qquad$ $\qquad$ gct :
* $\qquad$591198.87 $\qquad$$\qquad$$\qquad$$\qquad$ 591198.87
* $\qquad$328991.13 $\qquad$$\qquad$$\qquad$$\qquad$ 337809.30
* $\qquad$64598.674 $\qquad$$\qquad$$\qquad$$\qquad$ 65927.965
* $\qquad$12522.069 $\qquad$$\qquad$$\qquad$$\qquad$ -175.9230
* $\qquad$2336.1646 $\qquad$$\qquad$$\qquad$$\qquad$ -7.765696
* $\qquad$387.54032 $\qquad$$\qquad$$\qquad$$\qquad$ 1.5709228


4. D'après ce qui précède on remarque que l'algorithme de pas de cauchy permet d'avoir que des valeurs positives de la décroissance mais il est plus lent que celui de l'algorithme "gct", qui de son côté peut produire des valeurs négatives et donc augmenter la valeur de la fonction au lieu de la minimiser.


# Lagrangien augmenté

## Implémentation
1.Choisir des critères d’arrêt pour la convergence de l'algorithme.

2.Implémenter l'algorithme du lagrangien augmenté, en utilisant les différentes méthodes
qui ont été vues en première partie pour la résolution de la suite de problémes sans
contraintes (fichier `Lagrangien_Augmente.jl`)
 
3.Tester les différentes variantes sur les problèmes en Annexe D.
 

In [10]:
# Vos tests
include("Lagrangien_Augmente.jl")
include("Regions_De_Confiance.jl")
include("Gradient_Conjugue_Tronque.jl")
include("Pas_De_Cauchy.jl")

"Retour sur f1"
contrainte(x) = x[1]+x[3]-1
grad_contrainte(x) = [1; 0; 1]
hess_contrainte(x) = [0 0 0; 0 0 0; 0 0 0]

"réalisable"
xc = [0; 1; 1]
sol_exacte = []
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_gct","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_cauchy","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_newton","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)


"non réalisable"
xc = [0.5; 1.25; 1]
#xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
#my_afficher_resultats("Lagrangien_Augmente_gct","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_cauchy","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f1,contrainte,grad_f1,hess_f1,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_newton","f1",xc,xmin,f_min,flag,sol_exacte,nb_iters)




"Retour sur f2"
contrainte(x) =  (x[1]^2) + (x[2]^2) -1.5
grad_contrainte(x) = [2*x[1]; 2*x[2]]
hess_contrainte(x) = [2 0; 0 2]

"non réalisable"
xc = [1; 0]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_gct","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_cauchy","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_newton","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)

"réalisable"
xc = [sqrt(3)/2; sqrt(3)/2]
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("gct",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_gct","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("cauchy",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_cauchy","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)
xmin,f_min,flag,nb_iters = Lagrangien_Augmente("newton",f2,contrainte,grad_f2,hess_f2,grad_contrainte,hess_contrainte,xc,[])
my_afficher_resultats("Lagrangien_Augmente_newton","f2",xc,xmin,f_min,flag,sol_exacte,nb_iters)

-------------------------------------------------------------------------
Résultats de : Lagrangien_Augmente_gct appliqué à f1 au point initial [0, 1, 1]:
  * xsol = [0.5000002494183193, 1.2499998752908403, 0.5000002494183194]
  * f(xsol) = 2.249997755235685
  * nb_iters = 5
  * flag = 0
  * sol_exacte : Any[]
-------------------------------------------------------------------------
Résultats de : Lagrangien_Augmente_cauchy appliqué à f1 au point initial [0, 1, 1]:
  * xsol = [0.49999367482015405, 1.2499976023893034, 0.5000048471225219]
  * f(xsol) = 2.250006651355912
  * nb_iters = 5
  * flag = 0
  * sol_exacte : Any[]
-------------------------------------------------------------------------
Résultats de : Lagrangien_Augmente_newton appliqué à f1 au point initial [0, 1, 1]:
  * xsol = [0.5000002494183194, 1.2499998752908403, 0.5000002494183193]
  * f(xsol) = 2.249997755235685
  * nb_iters = 5
  * flag = 0
  * sol_exacte : Any[]
---------------------------------------------------------

## Interprétation
 1.Commenter les résultats obtenus, en étudiant notamment les valeurs de $\lambda_k$ et $\mu_k$.
 
 2.Étudier l'influence du paramètre $\tau$ dans la performance de l'algorithme.
 
 3.**Supplémentaire** : 
      Que proposez-vous comme méthode pour la résolution des problèmes avec
      des contraintes à la fois d'égalité et d'inégalité ? Implémenter (si le temps le permet)
      ce nouvel algorithme


## Vos réponses?
1.le nombre maximal d’iteration est de  1000 ,  $\lambda_0$  on lui a donné la valeur 2 et $\mu_0$ 100 et  $\tau$ 2 , le $\lambda$ ici n'a pas beaucoup d'influence sur notre algorithme, mais si $\lambda_0$ augmente on arrive à l'accélérer un peu.
Mais le $\lambda_k$ en sortie est important dans tous les cas ce qui nous amène à  dire  que la solution du problème sans contrainte ne satisfait pas aux contraintes qu'on applique.
Par contre  $\mu_k$ est imporatant dans tous les cas, car il est fortement lié aux nombres d'itérations qu'on réalise avant de retrouver la solution.Le $\mu_k$ facilite et permet de résoudre des problèmes complexes qui prennait beaucoup de temps avant. En augmentant $\mu_k$ le nombre d'itération effectuée diminue.le$\mu_{finale}$ augmente dans tous les cas. On peut expliquer cette augmentation par le non respect des contraintes qui nous oblige à ajouter des pénalités afin de les respecter.
cette algoritheme prend beaucoup de temps et cela car il faut réappliquer à chaque itération des algorithmes complexe car on modifie les facteurs  de contrainte . C'est un point négatif de cet algorithme.

2.On pourrait dire que plus $\tau$ est faible plus on a une meilleur convergence car on ne pénalise pas beaucoup mais c'est le contraire plus $\tau$ est élevé plus la convergence de l'algorithme est meilleur.
En choisant un critere d’arret sur la precision de la fonction, le résultat précis est notre priorité.
Mais il faut ici faire attention à choisir un valeur de tau qui soit adapté pas trop grande et pas trop petite car en choisissant une valeur de $\tau$ petite, on perd en précision, et si on choisit une valeur très grande on pourra nuire à la convergence.

3.La modification pour pouvoir résoudre des problèmes avec des contraintes à la fois d'égalité et d'inégalité et d'ajouter l'égalité et pour cela on pourra simplement exprimer une égalité comme deux égalité.Il faudra aussi voir le nouveau signe du gradient des contraintes.